scripts/generate_upstream_patch.sh

```diff
@@ -0,0 +1,138 @@
+#!/bin/bash
+
+set -eo pipefail
+
+cleanup() {
+  rm -rf "$1"
+  echo "cleaned up $1"
+}
+
+print_help() {
+  echo "Usage:"
+  echo "	$0 [-p <tag|branch|commit-hash>] [-w <tag|branch|commit-hash>]"
+  echo "	without any parameter, the script will generate upstream patch for both pallets and tee-worker"
+  echo "		-p specify the tag|branch|commit-hash for upstream pallets"
+  echo "		-w specify the tag|branch|commit-hash for upstream worker"
+}
+
+# This function generates a patch for the diffs between commit-A and commit-B
+# of the upstream repo, where
+# commit-A: the commit recorded in ./<TARGET_DIR>/upstream_commit
+# commit-B: the HEAD commit of upstream master or a given commit
+#
+# A patch will be generated under ./<TARGET_DIR>/upstream.patch
+generate_upstream_patch() {
+	local TARGET_DIR=$1
+	local UPSTREAM_NAME=$2
+	local UPSTREAM_URL=$3
+	# If $4 exists, it would be the target commit
+
+	cd $TARGET_DIR
+
+	if [ -f upstream_commit ]; then
+		OLD_COMMIT=$(head -1 upstream_commit)
+	else
+		echo "Can't find upstream_commit file in $TARGET_DIR, quit"
+		exit 1
+	fi
+
+	local tmp_dir=$(mktemp -d)
+	cd "$tmp_dir"
+	echo "cloning $UPSTREAM_URL to $tmp_dir"
+	git clone -q $UPSTREAM_URL repo
+	cd repo
+	[ "" != "$4" ] && git checkout "$4"
+	echo "generating $UPSTREAM_NAME.patch"
+	git diff $OLD_COMMIT HEAD > "$TARGET_DIR/$UPSTREAM_NAME.patch"
+	git rev-parse --short HEAD > "$TARGET_DIR/upstream_commit"
+	cleanup $tmp_dir
+	echo
+}
+
+if [[ "-h" == "$1" ]]
+then
+	print_help
+	exit 1
+fi
+
+while getopts ":p:w:h:" opt; do
+	case $opt in
+		p)
+			has_pallets=true
+			pallets_commit=$OPTARG
+			;;
+		w)
+			HAS_WORKER=true
+			WORKER_COMMIT=$OPTARG
```

Comment:
Mixed big and small letters seem confusing, is it intended or wrong?

I tried to run `./scripts/generate_upstream_patch.sh -p`, then both pallets and worker patches were generated. Is it intended?

Reply:
-p must follow a parameter; otherwise it will be dropped off.